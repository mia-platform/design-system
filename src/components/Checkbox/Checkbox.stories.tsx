/**
 * Copyright 2024 Mia srl
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * SPDX-License-Identifier: Apache-2.0
 */

import { Meta, StoryObj } from '@storybook/react'

import { Checkbox } from '.'
import { defaults } from './Checkbox'

const meta = {
  component: Checkbox,
  args: {
    ...defaults,
  },
} satisfies Meta<typeof Checkbox>

export default meta
type Story = StoryObj<typeof meta>

export const BasicExample: Story = {}

export const CheckedOnFirstRender: Story = {
  args: {
    isInitiallyChecked: true,
  },
}

export const ControlledCheckedState: Story = {
  args: {
    isChecked: true,
  },
}

export const Indeterminate: Story = {
  args: {
    isIndeterminate: true,
  },
}

export const Disabled: Story = {
  args: {
    isDisabled: true,
  },
}

export const IndeterminateDisabled: Story = {
  args: {
    isDisabled: true,
    isIndeterminate: true,
  },
}

export const CheckedDisabled: Story = {
  args: {
    isChecked: true,
    isDisabled: true,
  },
}

export const WithText: Story = {
  args: {
    label: 'Checkbox with text',
  },
}

export const WithTextAndDescription: Story = {
  args: {
    label: 'Checkbox with text',
    description: 'This is a description of the switch.',
  },
}

export const DisabledWithText: Story = {
  args: {
    isDisabled: true,
    label: 'Checkbox with text',
  },
}

export const DisabledWithTextAndDescription: Story = {
  args: {
    isDisabled: true,
    label: 'Checkbox with text',
    description: 'This is a description of the switch.',
  },
}

export const WithOnChange: Story = {
  args: {
    onChange: (checked) => alert(`Checkbox is now ${checked ? 'checked' : 'unchecked'}`),
  },
}
