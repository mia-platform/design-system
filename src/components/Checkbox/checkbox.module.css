/**
 * Copyright 2024 Mia srl
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * SPDX-License-Identifier: Apache-2.0
 */

.checkbox {
  margin-inline: 0;
  align-items: flex-start;
  line-height: 1rem;

  & .checkboxContent {
    display: flex;
    flex-direction: column;
    gap: var(--spacing-gap-sm, 8px);
  }

  &:global(.mia-platform-checkbox-wrapper) {
    display: flex;
    align-items: center;
  }

  &:not(:global(.mia-platform-checkbox-wrapper-disabled)) {
    &:hover :global(.mia-platform-checkbox-inner) {
      border-color: var(--palette-action-primary-hover, #1890FF) !important;
    }

    &  :global(.mia-platform-checkbox-indeterminate) :global(.mia-platform-checkbox-inner)::after {
      border-radius: var(--shape-border-radius-xs, -1px);
    }

    &:hover :global(.mia-platform-checkbox-indeterminate) :global(.mia-platform-checkbox-inner)::after {
      background-color: var(--palette-action-primary-hover, #1890FF);
    }
  }

  /* disable */
  &:global(.mia-platform-checkbox-wrapper-disabled) {
    :global(.mia-platform-typography) {
      color: var(--palette-text-neutral-subtler, #ACACAC);
    }
    & :global(.mia-platform-checkbox-inner) {
      background-color: var(--palette-action-disabled-main, #F2F2F2) !important;
      border-color: var(--palette-action-disabled-bolder, #CDCDCD) !important;
      &:after {
        border-color: var(--palette-action-disabled-contrastText, #ACACAC);
      }
    }

    & :global(.mia-platform-checkbox-indeterminate) :global(.mia-platform-checkbox-inner)::after {
      background-color: var(--palette-action-disabled-contrastText, #ACACAC);
    }
  }

  & :global(span.mia-platform-checkbox) {
    align-self: unset !important;
    & + * {
      padding-inline-start: var(--spacing-gap-sm, 8px);
      padding-inline-end: var(--spacing-gap-sm, 8px);
    }
  }
}
