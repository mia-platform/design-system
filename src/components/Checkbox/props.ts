/**
 * Copyright 2024 Mia srl
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * SPDX-License-Identifier: Apache-2.0
 */

import { CheckboxChangeEvent } from 'antd/es/checkbox'
import { ReactNode } from 'react'

export type CheckboxProps = {

  /**
   * Additional description of the checkbox, to be rendered below the checkbox `text`.
   * Is ignored if the `text` prop is not provided.
   */
  description?: ReactNode,

  /**
   * Allows you to control whether the checkbox is checked.
   */
  isChecked?: boolean,

  /**
   * Allows you to control whether the checkbox is disabled. Defaults to false.
   */
  isDisabled?: boolean,

  /**
   * Allows you to control whether the checkbox is checked on its first render. Defaults to false.
   */
  isInitiallyChecked?: boolean,

  /**
   * Allows you to control whether the checkbox is indeterminate. Defaults to false.
   */
  isIndeterminate?: boolean

  /**
   * Function that is invoked when the checkbox state is changed.
   */
  onChange?: (event: CheckboxChangeEvent) => void

  /**
   * Text to be displayed next to the checkbox.
   */
  text?: ReactNode
}
