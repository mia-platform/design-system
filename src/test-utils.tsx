/**
 * Copyright 2023 Mia srl
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * SPDX-License-Identifier: Apache-2.0
 */

import { RenderOptions, RenderResult, render } from '@testing-library/react'
import { ConfigProvider } from 'antd'
import { ReactNode } from 'react'

import { ThemeContext, defaultTheme } from './components/ThemeProvider'
import { ANT_PREFIX } from './components/ThemeProvider/Ant'

/**
 * Wraps a React node with a series of providers for testing purposes.
 *
 * @remarks It adds the default theme to components
 * @remarks It adds the default Ant class prefix
 * @remarks It removes the css-do-not-override classes from Ant Design components
 *
 * @param {ReactNode} children - The React node to be wrapped.
 * @returns {ReactNode} The wrapped React node.
 */
const TestWrapper = ({ children }: {children: ReactNode}): ReactNode => (
  <ThemeContext.Provider value={defaultTheme}>
    <ConfigProvider prefixCls={ANT_PREFIX} theme={{ hashed: false }}>
      {children}
    </ConfigProvider>
  </ThemeContext.Provider>
)

/**
 * Custom render function for testing React components.
 *
 * @param {ReactNode} element - The React node to be rendered.
 * @param {RenderOptions} [options] - Options for rendering.
 * @returns {RenderResult} The result of rendering the React node.
 */
const customRender = (
  element: ReactNode,
  options?: Omit<RenderOptions, 'wrapper'>,
): RenderResult => render(element, { wrapper: TestWrapper, ...options })

export * from '@testing-library/react'
export {
  customRender as render,
  TestWrapper,
}
