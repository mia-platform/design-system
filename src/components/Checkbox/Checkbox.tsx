/**
 * Copyright 2024 Mia srl
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * SPDX-License-Identifier: Apache-2.0
 */

import { ReactElement, useMemo } from 'react'
import { Checkbox as AntCheckbox } from 'antd'
import classnames from 'classnames'

import { CheckboxProps } from './props'
import styles from './checkbox.module.css'

export const defaults = {
  isDisabled: false,
  isInitiallyChecked: false,
  isIndeterminate: false,
}

export const Checkbox = <T = never, >({
  label,
  description,
  isInitiallyChecked = defaults.isInitiallyChecked,
  isIndeterminate = defaults.isIndeterminate,
  isDisabled = defaults.isDisabled,
  onChange,
  isChecked,
  value,
}: CheckboxProps<T>): ReactElement => {
  const className = useMemo(() => classnames([
    styles.checkboxComponent,
    isDisabled && styles.disabled,
  ]), [isDisabled])

  return (
    <div className={className}>
      <div className={styles.checkboxTextWrapper}>
        <AntCheckbox
          defaultChecked={isInitiallyChecked}
          disabled={isDisabled}
          indeterminate={isIndeterminate}
          onChange={onChange}
          {...(isChecked !== undefined && { checked: isChecked })}
          {...(value !== undefined && { value })}
        />
        {label}
      </div>
      {
        description && (
          <div className={styles.checkboxDescription}>
            {description}
          </div>
        )
      }
    </div>
  )
}
