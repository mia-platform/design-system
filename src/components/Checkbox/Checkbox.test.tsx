/**
 * Copyright 2024 Mia srl
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * SPDX-License-Identifier: Apache-2.0
 */

import { render, screen, userEvent, waitFor } from '../../test-utils'
import { Checkbox } from './Checkbox'

describe('Checkbox', () => {
  describe('renders states correctly', () => {
    it('renders default state', () => {
      render(<Checkbox />)
      expect(screen.getByRole('checkbox')).toBeInTheDocument()
      expect(screen.getByRole('checkbox')).not.toBeChecked()
    })

    it('renders checked by default state', () => {
      render(<Checkbox isInitiallyChecked={true} />)
      expect(screen.getByRole('checkbox')).toBeInTheDocument()
      expect(screen.getByRole('checkbox')).toBeChecked()
    })

    it('renders checked state', () => {
      render(<Checkbox isChecked={true} />)
      expect(screen.getByRole('checkbox')).toBeInTheDocument()
      expect(screen.getByRole('checkbox')).toBeChecked()
    })

    it('renders default disabled state', () => {
      render(<Checkbox isDisabled={true} />)
      expect(screen.getByRole('checkbox')).toBeInTheDocument()
      expect(screen.getByRole('checkbox')).toBeDisabled()
    })

    it('renders checked disabled state', () => {
      const { asFragment } = render(<Checkbox isChecked={true} isDisabled={true} />)
      expect(screen.getByRole('checkbox')).toBeInTheDocument()
      expect(screen.getByRole('checkbox')).toBeDisabled()
      expect(screen.getByRole('checkbox')).toBeChecked()
      expect(asFragment()).toMatchSnapshot()
    })
  })

  describe('renders text correctly', () => {
    it('renders simple text', () => {
      render(<Checkbox label="Checkbox" />)
      expect(screen.getByRole('checkbox')).toBeInTheDocument()
      expect(screen.getByText('Checkbox')).toBeInTheDocument()
    })

    it('renders text and description', () => {
      render(<Checkbox description="Checkbox description" label="Checkbox text" />)
      expect(screen.getByRole('checkbox')).toBeInTheDocument()
      expect(screen.getByText('Checkbox text')).toBeInTheDocument()
      expect(screen.getByText('Checkbox description')).toBeInTheDocument()
    })

    it('ignores description if text is not set', () => {
      render(<Checkbox description="Checkbox description" />)
      expect(screen.getByRole('checkbox')).toBeInTheDocument()
      expect(screen.queryByText('Checkbox description')).not.toBeInTheDocument()
    })
  })

  describe('performs interactions correctly', () => {
    it('calls onChange when the checkbox state changes', async() => {
      const onChange = jest.fn()
      render(<Checkbox onChange={onChange} />)
      expect(screen.getByRole('checkbox')).toBeInTheDocument()
      expect(screen.getByRole('checkbox')).not.toBeChecked()
      await userEvent.click(screen.getByRole('checkbox'))
      await waitFor(() => expect(onChange).toHaveBeenCalledTimes(1))
    })
  })
})
