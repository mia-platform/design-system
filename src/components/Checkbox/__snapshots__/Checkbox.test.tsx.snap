// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Checkbox renders states correctly renders checked disabled state 1`] = `
<DocumentFragment>
  <label
    class="mia-platform-checkbox-wrapper mia-platform-checkbox-wrapper-checked mia-platform-checkbox-wrapper-disabled checkbox"
  >
    <span
      class="mia-platform-checkbox ant-wave-target mia-platform-checkbox-checked mia-platform-checkbox-disabled"
    >
      <input
        checked=""
        class="mia-platform-checkbox-input"
        disabled=""
        type="checkbox"
      />
      <span
        class="mia-platform-checkbox-inner"
      />
    </span>
    <span>
      <div
        class="checkboxContent"
      />
    </span>
  </label>
</DocumentFragment>
`;
